src/main/java/com/google/cloud/tools/gradle/appengine/core/CloudSdkDownloader.java

```diff
@@ -17,11 +17,199 @@
+import org.gradle.api.logging.Logger;
 
-interface CloudSdkDownloader {
+public class CloudSdkDownloader {
 
-  File downloadSdk(String version);
+  private Logger logger;
 
-  boolean isSdkValid(String version, File home);
+  private SdkInstaller installer;
+  private SdkComponentInstaller componentInstaller;
+  private SdkUpdater updater;
+
+  public CloudSdkDownloader() {
+    this.logger = null;
+  }
+
+  public CloudSdkDownloader(Logger logger) {
+    this.logger = logger;
+  }
+
+  public void setInstaller(SdkInstaller installer) {
+    this.installer = installer;
+  }
+
+  public void setComponentInstaller(SdkComponentInstaller componentInstaller) {
+    this.componentInstaller = componentInstaller;
+  }
+
+  public void setUpdater(SdkUpdater updater) {
+    this.updater = updater;
+  }
+
+  /**
+   * Download the SDK at specified version, or update existing version.
+   *
+   * @return the downloaded SDK's home directory.
+   * @throws GradleException when installation fails.
+   */
+  public File downloadSdk(String version) {
+    // Create ManagedCloudSdk from given version
+    ManagedCloudSdk sdk;
+    try {
+      if (version.equals(Version.LATEST.getVersion())) {
+        sdk = ManagedCloudSdk.newManagedSdk();
+      } else {
+        sdk = ManagedCloudSdk.newManagedSdk(new Version(version));
+      }
+    } catch (UnsupportedOsException | BadCloudSdkVersionException e) {
+      throw new GradleException(e.getMessage());
+    }
+
+    ProgressListener progressListener =
+        new ProgressListener() {
+          @Override
+          public void start(String message, long totalWork) {
+            if (logger != null) {
+              logger.lifecycle(message);
+            }
+          }
+
+          @Override
+          public void update(long workDone) {
+            // TODO: Show progress
+          }
+
+          @Override
+          public void update(String message) {
+            if (logger != null) {
+              logger.lifecycle(message);
+            }
+          }
+
+          @Override
+          public void done() {}
+
+          @Override
+          public ProgressListener newChild(long allocation) {
+            return new ProgressListener() {
+              @Override
+              public void start(String message, long totalWork) {}
+
+              @Override
+              public void update(long workDone) {}
+
+              @Override
+              public void update(String message) {}
+
+              @Override
+              public void done() {}
+
+              @Override
+              public ProgressListener newChild(long allocation) {
+                return null;
+              }
+            };
+          }
+        };
+
+    ConsoleListener consoleListener =
+        new ConsoleListener() {
+          @Override
+          public void console(String rawString) {}
+        };
+
+    try {
+      // Install sdk if not installed
+      if (!sdk.isInstalled()) {
+        if (installer == null) {
```

Comment:
This this `null`-check could potentially cause a bug if you re-use the `downloadSdk` method, but the actual `ManagedCloudSdk` used is different (should have created a new `Installer` but used an old one).

This `null`-check seems to be used for just testing purposes, where you use a mocked `Installer` that gets set as the `installer` field. It's generally a good idea not to have logic in your code that is only for the purpose of testing (in other words, generally, [tests should be blind to the actual implementation of the method they are testing](https://testing.googleblog.com/2013/08/testing-on-toilet-test-behavior-not.html) - method call verifications violate this and should really be avoided unless necessary).

A possible solution here might be to mock the `ManagedCloudSdk#newInstaller` and return your mocked `Installer`.